File: orion-platform-service/src/services/chatops/WebhookService.ts
/**
 * Webhook Service
 *
 * Manages ChatOps webhook configurations and delivery
 */

import { DatabasePool } from '../database';
import { v4 as uuidv4 } from 'uuid';
import crypto from 'crypto';

export interface WebhookConfig {
  id: string;
  name: string;
  url: string;
  events: string[];
  secret_key: string | null;
  enabled: boolean;
  retry_count: number;
  retry_interval_seconds: number;
  timeout_seconds: number;
  headers: Record<string, string>;
  description: string;
  created_by: string;
  last_triggered_at: Date | null;
  last_status: string | null;
  created_at: Date;
  updated_at: Date;
}

export interface CreateWebhookInput {
  name: string;
  url: string;
  events: string[];
  secret_key?: string;
  enabled?: boolean;
  retry_count?: number;
  retry_interval_seconds?: number;
  timeout_seconds?: number;
  headers?: Record<string, string>;
  description?: string;
  created_by?: string;
}

export interface UpdateWebhookInput {
  name?: string;
  url?: string;
  events?: string[];
  secret_key?: string;
  enabled?: boolean;
  retry_count?: number;
  retry_interval_seconds?: number;
  timeout_seconds?: number;
  headers?: Record<string, string>;
  description?: string;
}

export interface WebhookLog {
  id: string;
  webhook_id: string;
  event_type: string;
  payload: Record<string, unknown>;
  response_status: number | null;
  response_body: string | null;
  error_message: string | null;
  retry_count: number;
  created_at: Date;
}

export class WebhookService {
  constructor(private pool: DatabasePool) {}

  async getAll(): Promise<WebhookConfig[]> {
    const result = await this.pool.query(
      'SELECT * FROM chatops_webhooks ORDER BY created_at DESC'
    );
    return result.rows.map(row => ({
      ...row,
      events: Array.isArray(row.events) ? row.events : JSON.parse(row.events || '[]'),
      headers: typeof row.headers === 'string' ? JSON.parse(row.headers) : row.headers,
    }));
  }

  async getById(id: string): Promise<WebhookConfig | null> {
    const result = await this.pool.query(
      'SELECT * FROM chatops_webhooks WHERE id = $1',
      [id]
    );
    if (!result.rows[0]) return null;
    const row = result.rows[0];
    return {
      ...row,
      events: Array.isArray(row.events) ? row.events : JSON.parse(row.events || '[]'),
      headers: typeof row.headers === 'string' ? JSON.parse(row.headers) : row.headers,
    };
  }

  async create(input: CreateWebhookInput): Promise<WebhookConfig> {
    const id = uuidv4();
    const now = new Date();
    const secret = input.secret_key || crypto.randomBytes(32).toString('hex');

    await this.pool.query(
      `INSERT INTO chatops_webhooks
       (id, name, url, events, secret_key, enabled, retry_count, retry_interval_seconds, timeout_seconds, headers, description, created_by, created_at, updated_at)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14)`,
      [id, input.name, input.url, JSON.stringify(input.events), secret,
       input.enabled ?? true, input.retry_count ?? 3, input.retry_interval_seconds ?? 30,
       input.timeout_seconds ?? 10, JSON.stringify(input.headers || {}),
       input.description || '', input.created_by || 'system', now, now]
    );

    return this.getById(id) as Promise<WebhookConfig>;
  }

  async update(id: string, input: UpdateWebhookInput): Promise<WebhookConfig | null> {
    const existing = await this.getById(id);
    if (!existing) return null;

    const updates: string[] = [];
    const params: any[] = [];
    let pi = 1;

    if (input.name !== undefined) { updates.push(`name = $${pi++}`); params.push(input.name); }
    if (input.url !== undefined) { updates.push(`url = $${pi++}`); params.push(input.url); }
    if (input.events !== undefined) { updates.push(`events = $${pi++}`); params.push(JSON.stringify(input.events)); }
    if (input.secret_key !== undefined) { updates.push(`secret_key = $${pi++}`); params.push(input.secret_key || null); }
    if (input.enabled !== undefined) { updates.push(`enabled = $${pi++}`); params.push(input.enabled); }
    if (input.retry_count !== undefined) { updates.push(`retry_count = $${pi++}`); params.push(input.retry_count); }
    if (input.retry_interval_seconds !== undefined) { updates.push(`retry_interval_seconds = $${pi++}`); params.push(input.retry_interval_seconds); }
    if (input.timeout_seconds !== undefined) { updates.push(`timeout_seconds = $${pi++}`); params.push(input.timeout_seconds); }
    if (input.headers !== undefined) { updates.push(`headers = $${pi++}`); params.push(JSON.stringify(input.headers)); }
    if (input.description !== undefined) { updates.push(`description = $${pi++}`); params.push(input.description); }

    if (updates.length > 0) {
      updates.push(`updated_at = $${pi++}`);
      params.push(new Date(), id);
      await this.pool.query(
        `UPDATE chatops_webhooks SET ${updates.join(', ')} WHERE id = $${pi}`,
        params
      );
    }

    return this.getById(id);
  }

  async delete(id: string): Promise<boolean> {
    const result = await this.pool.query('DELETE FROM chatops_webhooks WHERE id = $1', [id]);
    return (result.rowCount ?? 0) > 0;
  }

  async testWebhook(id: string): Promise<{ success: boolean; status?: number; error?: string }> {
    const webhook = await this.getById(id);
    if (!webhook) return { success: false, error: 'Webhook not found' };

    try {
      // Simulate delivery test
      const response = await fetch(webhook.url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-ChatOps-Signature': this.signPayload(JSON.stringify({ test: true }), webhook.secret_key),
          ...(webhook.headers || {}),
        },
        body: JSON.stringify({ event: 'test', timestamp: new Date().toISOString() }),
        signal: AbortSignal.timeout((webhook.timeout_seconds || 10) * 1000),
      });

      await this.logDelivery(id, 'test', { test: true }, response.status, await response.text());
      return { success: response.ok, status: response.status };
    } catch (err) {
      const errorMsg = err instanceof Error ? err.message : 'Unknown error';
      await this.logDelivery(id, 'test', { test: true }, null, errorMsg);
      return { success: false, error: errorMsg };
    }
  }

  async getLogs(webhookId: string, limit = 20): Promise<WebhookLog[]> {
    const result = await this.pool.query(
      `SELECT * FROM chatops_webhook_logs
       WHERE webhook_id = $1
       ORDER BY created_at DESC
       LIMIT $2`,
      [webhookId, limit]
    );
    return result.rows.map(row => ({
      ...row,
      payload: typeof row.payload === 'string' ? JSON.parse(row.payload) : row.payload,
    }));
  }

  private async logDelivery(
    webhookId: string,
    eventType: string,
    payload: Record<string, unknown>,
    status: number | null,
    responseBody: string | null,
    errorMessage?: string,
    retryCount = 0
  ): Promise<void> {
    await this.pool.query(
      `INSERT INTO chatops_webhook_logs
       (id, webhook_id, event_type, payload, response_status, response_body, error_message, retry_count)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
      [uuidv4(), webhookId, eventType, JSON.stringify(payload),
       status, responseBody, errorMessage || null, retryCount]
    );

    // Update webhook last_triggered_at and last_status
    await this.pool.query(
      `UPDATE chatops_webhooks
       SET last_triggered_at = NOW(), last_status = $1
       WHERE id = $2`,
      [errorMessage ? 'failed' : (status === 200 ? 'success' : 'error'), webhookId]
    );
  }

  private signPayload(payload: string, secret: string | null): string {
    if (!secret) return '';
    return crypto.createHmac('sha256', secret).update(payload).digest('hex');
  }

  /**
   * Deliver an event to matching webhooks
   */
  async deliverEvent(eventType: string, payload: Record<string, unknown>): Promise<void> {
    const webhooks = await this.getAll();
    const matching = webhooks.filter(w => w.enabled && w.events.includes(eventType));

    for (const webhook of matching) {
      try {
        const response = await fetch(webhook.url, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'X-ChatOps-Signature': this.signPayload(JSON.stringify(payload), webhook.secret_key),
            ...(webhook.headers || {}),
          },
          body: JSON.stringify({ event: eventType, ...payload, timestamp: new Date().toISOString() }),
          signal: AbortSignal.timeout((webhook.timeout_seconds || 10) * 1000),
        });

        await this.logDelivery(webhook.id, eventType, payload, response.status, await response.text());
      } catch (err) {
        const errorMsg = err instanceof Error ? err.message : 'Unknown error';
        await this.logDelivery(webhook.id, eventType, payload, null, errorMsg);
      }
    }
  }
}

File: orion-chatops-svc-go/internal/service/webhook_service.go
package service

import (
	"bytes"
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"sync"
	"time"

	"orion/chatops-svc-go/internal/models"
	"orion/chatops-svc-go/internal/repository"

	"github.com/google/uuid"
)

// WebhookService handles webhook CRUD and event delivery.
type WebhookService struct {
	repo   *repository.Repository
	client *http.Client
	wg     sync.WaitGroup
	cancel context.CancelFunc
}

func NewWebhookService(repo *repository.Repository) *WebhookService {
	ctx, cancel := context.WithCancel(context.Background())
	_ = ctx // ctx is used in Shutdown
	return &WebhookService{
		repo:   repo,
		client: &http.Client{Timeout: 30 * time.Second},
		cancel: cancel,
	}
}

// Shutdown cancels all in-flight deliveries and waits for them to finish.
func (s *WebhookService) Shutdown() {
	s.cancel()
	s.wg.Wait()
}

func (s *WebhookService) Create(ctx context.Context, tenantID string, req models.CreateWebhookRequest) (*models.ChatOpsWebhook, error) {
	enabled := true
	if req.Enabled != nil {
		enabled = *req.Enabled
	}
	retryCount := req.RetryCount
	if retryCount <= 0 {
		retryCount = 3
	}
	retryInterval := req.RetryIntervalSeconds
	if retryInterval <= 0 {
		retryInterval = 5
	}
	timeout := req.TimeoutSeconds
	if timeout <= 0 {
		timeout = 30
	}

	h := &models.ChatOpsWebhook{
		ID:                   uuid.New().String(),
		TenantID:             tenantID,
		Name:                 req.Name,
		URL:                  req.URL,
		Events:               req.Events,
		Enabled:              enabled,
		RetryCount:           retryCount,
		RetryIntervalSeconds: retryInterval,
		TimeoutSeconds:       timeout,
		Description:          req.Description,
		CreatedBy:            req.CreatedBy,
	}
	if req.SecretKey != "" {
		h.SecretKey = &req.SecretKey
	}
	if req.Headers != nil {
		h.Headers = models.JSONB{}
		for k, v := range req.Headers {
			h.Headers[k] = v
		}
	}
	if err := s.repo.CreateWebhook(ctx, h); err != nil {
		return nil, err
	}
	return h, nil
}

func (s *WebhookService) Get(ctx context.Context, tenantID, id string) (*models.ChatOpsWebhook, error) {
	return s.repo.GetWebhook(ctx, tenantID, id)
}

func (s *WebhookService) List(ctx context.Context, tenantID string) ([]models.ChatOpsWebhook, error) {
	return s.repo.ListWebhooks(ctx, tenantID)
}

func (s *WebhookService) Update(ctx context.Context, tenantID, id string, req models.UpdateWebhookRequest) (*models.ChatOpsWebhook, error) {
	existing, err := s.repo.GetWebhook(ctx, tenantID, id)
	if err != nil {
		return nil, err
	}
	if req.Name != nil {
		existing.Name = *req.Name
	}
	if req.URL != nil {
		existing.URL = *req.URL
	}
	if req.Events != nil {
		existing.Events = *req.Events
	}
	if req.SecretKey != nil {
		existing.SecretKey = req.SecretKey
	}
	if req.Enabled != nil {
		existing.Enabled = *req.Enabled
	}
	if req.RetryCount != nil {
		existing.RetryCount = *req.RetryCount
	}
	if req.RetryIntervalSeconds != nil {
		existing.RetryIntervalSeconds = *req.RetryIntervalSeconds
	}
	if req.TimeoutSeconds != nil {
		existing.TimeoutSeconds = *req.TimeoutSeconds
	}
	if req.Headers != nil {
		existing.Headers = models.JSONB{}
		for k, v := range req.Headers {
			existing.Headers[k] = v
		}
	}
	if req.Description != nil {
		existing.Description = *req.Description
	}
	if err := s.repo.UpdateWebhook(ctx, existing); err != nil {
		return nil, err
	}
	return existing, nil
}

func (s *WebhookService) Delete(ctx context.Context, tenantID, id string) error {
	return s.repo.DeleteWebhook(ctx, tenantID, id)
}

func (s *WebhookService) GetLogs(ctx context.Context, tenantID, webhookID string, limit int) ([]models.ChatOpsWebhookLog, error) {
	if limit <= 0 {
		limit = 50
	}
	return s.repo.ListWebhookLogs(ctx, tenantID, webhookID, limit)
}

// DeliverEvent sends an event to all enabled webhooks subscribed to the event type.
func (s *WebhookService) DeliverEvent(ctx context.Context, tenantID, eventType string, payload interface{}) error {
	webhooks, err := s.repo.ListEnabledWebhooks(ctx, tenantID)
	if err != nil {
		return err
	}

	for _, wh := range webhooks {
		if !s.eventMatches(wh.Events, eventType) {
			continue
		}
		s.wg.Add(1)
		go func(w models.ChatOpsWebhook) {
			defer s.wg.Done()
			s.deliverWithRetry(w, eventType, payload)
		}(wh)
	}
	return nil
}

func (s *WebhookService) eventMatches(events []string, eventType string) bool {
	for _, e := range events {
		if e == "*" || e == eventType {
			return true
		}
	}
	return false
}

func (s *WebhookService) deliverWithRetry(wh models.ChatOpsWebhook, eventType string, payload interface{}) {
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(wh.RetryCount+1)*time.Duration(wh.RetryIntervalSeconds)*time.Second+time.Duration(wh.TimeoutSeconds)*time.Second)
	defer cancel()

	body, err := json.Marshal(map[string]interface{}{
		"event":     eventType,
		"payload":   payload,
		"timestamp": time.Now().UTC().Format(time.RFC3339),
	})
	if err != nil {
		log.Printf("webhook: failed to marshal payload for %s: %v", wh.ID, err)
		return
	}

	var lastErr string
	var lastStatus int
	for attempt := 0; attempt <= wh.RetryCount; attempt++ {
		if attempt > 0 {
			timer := time.NewTimer(time.Duration(wh.RetryIntervalSeconds) * time.Second)
			select {
			case <-timer.C:
				// continue with retry
			case <-ctx.Done():
				timer.Stop()
				return
			}
		}

		req, err := http.NewRequestWithContext(ctx, "POST", wh.URL, bytes.NewReader(body))
		if err != nil {
			log.Printf("webhook: failed to create request for %s: %v", wh.ID, err)
			return
		}
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("X-Orion-Event", eventType)

		// HMAC signature if secret exists
		if wh.SecretKey != nil && *wh.SecretKey != "" {
			sig := computeHMAC(body, *wh.SecretKey)
			req.Header.Set("X-Orion-Signature", sig)
		}

		// Custom headers
		if wh.Headers != nil {
			for k, v := range wh.Headers {
				if vs, ok := v.(string); ok {
					req.Header.Set(k, vs)
				}
			}
		}

		resp, err := s.client.Do(req)
		if err != nil {
			lastErr = err.Error()
			continue
		}
		defer resp.Body.Close()
		respBody, _ := io.ReadAll(resp.Body)

		lastStatus = resp.StatusCode
		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
			// Success
			if err := s.repo.UpdateWebhookTriggerStatus(ctx, wh.TenantID, wh.ID, "success"); err != nil {
				log.Printf("webhook: failed to update trigger status for %s: %v", wh.ID, err)
			}
			if err := s.repo.CreateWebhookLog(ctx, &models.ChatOpsWebhookLog{
				ID:             uuid.New().String(),
				TenantID:       wh.TenantID,
				WebhookID:      wh.ID,
				EventType:      eventType,
				Payload:        models.JSONB{"body": string(body)},
				ResponseStatus: &resp.StatusCode,
				ResponseBody:   strPtr(string(respBody)),
				RetryCount:     attempt,
			}); err != nil {
				log.Printf("webhook: failed to create log for %s: %v", wh.ID, err)
			}
			return
		}
		lastErr = fmt.Sprintf("HTTP %d: %s", resp.StatusCode, string(respBody))
	}

	// All retries exhausted
	if err := s.repo.UpdateWebhookTriggerStatus(ctx, wh.TenantID, wh.ID, "failed"); err != nil {
		log.Printf("webhook: failed to update trigger status for %s: %v", wh.ID, err)
	}
	if err := s.repo.CreateWebhookLog(ctx, &models.ChatOpsWebhookLog{
		ID:             uuid.New().String(),
		TenantID:       wh.TenantID,
		WebhookID:      wh.ID,
		EventType:      eventType,
		Payload:        models.JSONB{"body": string(body)},
		ResponseStatus: &lastStatus,
		ErrorMessage:   &lastErr,
		RetryCount:     wh.RetryCount,
	}); err != nil {
		log.Printf("webhook: failed to create log for %s: %v", wh.ID, err)
	}
}

func computeHMAC(body []byte, secret string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(body)
	return "sha256=" + hex.EncodeToString(mac.Sum(nil))
}

func strPtr(s string) *string { return &s }

// Test sends a test payload to a webhook and returns the result.
func (s *WebhookService) Test(ctx context.Context, tenantID, id string) (map[string]interface{}, error) {
	wh, err := s.repo.GetWebhook(ctx, tenantID, id)
	if err != nil {
		return nil, err
	}

	testPayload := map[string]interface{}{
		"event":     "test",
		"timestamp": time.Now().UTC().Format(time.RFC3339),
		"payload":   map[string]string{"message": "This is a test webhook notification"},
	}

	body, _ := json.Marshal(testPayload)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, wh.URL, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Webhook-Event", "test")
	if wh.SecretKey != nil {
		req.Header.Set("X-Signature-256", computeHMAC(body, *wh.SecretKey))
	}

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return map[string]interface{}{
			"success": false,
			"error":   err.Error(),
		}, nil
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)
	return map[string]interface{}{
		"success":    resp.StatusCode >= 200 && resp.StatusCode < 300,
		"statusCode": resp.StatusCode,
		"response":   string(respBody),
	}, nil
}
